lib/rubocop/cop/rspec/instance_spy.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks for `instance_double(Foo).as_null_object usage`
+      # when the double is used with `have_received`.
+      #
+      # @example
+      # # bad
+      # it do
+      #  foo = instance_double(Foo).as_null_object
+      #  expect(foo).to have_received(:bar)
+      # end
+      #
+      # # good
+      # it do
+      #   foo = instance_spy(Foo)
+      #   expect(foo).to have_received(:bar)
+      # end
+      #
+      class InstanceSpy < Cop
+        MSG = 'Use `instance_spy` when you check your double ' \
+              'with `have_received`'.freeze
+
+        EXAMPLE_BLOCKS = RuboCop::RSpec::Language::Examples::ALL
+
+        def_node_search :spy, <<-PATTERN
+        (lvasgn $_
+          (send
+            $(send nil :instance_double
+              (const nil _)) :as_null_object))
+        PATTERN
+
+        def_node_search :expectation, <<-PATTERN
+        (send
+          (send nil :expect
+          (lvar $_)) :to
+          (send nil :have_received
+          ...)
+        ...)
+        PATTERN
+
+        def on_block(node)
+          method, _args, _body = *node
+          _receiver, method_name, _object = *method
+          return unless EXAMPLE_BLOCKS.include?(method_name)
+
+          spy(node) do |var, receiver|
+            expectation(node) do |expected|
+              add_offense(receiver, :expression, MSG) if expected == var
+            end
+          end
+        end
+
+        def autocorrect(node)
+          # node is instance_double(Foo)
+          _receiver, _method, arg = *node
+          _type, receiver, _args = *arg
+
+          replacement = "instance_spy(#{receiver})"
```

Comment:
You probably want to figure out how to just rename the selector without building the source manually like this. This will make it easier to handle cases too where we have `instance_double(Foo, a: :b)`